SwiftCompilerSources/Sources/SIL/FunctionConvention.swift

```diff
@@ -164,7 +164,7 @@ public struct ParameterInfo : CustomStringConvertible {
     switch convention {
     case .indirectIn, .indirectInGuaranteed:
       return hasLoweredAddresses || type.isOpenedExistentialWithError()
-    case .indirectInout, .indirectInoutAliasable:
+    case .indirectInout, .indirectInoutAliasable, .indirectInCXX:
```

Comment:
cc @atrick

Reply:
It's potentially useful to have a SIL convention that says: the caller and callee must observe the same address of an addressable value. But that isn't generally implementable for all `T` given reabstraction of function types and tuples.

I'm still baffled by what this convention means. The comment says it's about observing mutation. If that's true, then we should pass a SIL address in non-lowered-addresses mode. But why, then aren't you using `inout`?

If it is *not* the case that the caller and callee need to observe the same address, then we should pass SIL values instead.

Reply:
I saw an assertion failure when I tried using `inout` by making `getIndirectCParameterConvention` return `Indirect_Inout` instead of `Indirect_In_CXX`.

The failed assertion is in `ArgEmitter::emit`:

```
assert(arg.hasLValueType() == param.isIndirectInOut());
```

Reply:
This is the Itanium C++ ABI convention in which non-trivial C++ values are passed indirectly but the caller is responsible for destroying them.  The caller and callee do need to observe the same address.  It is essentially the same as `inout` at the SIL level except that it's not `inout` in the source language, and I didn't want us to have to weaken all the SIL lowering assertions that e.g. an `inout` parameter always corresponds to an l-value argument expression, because those are quite useful in catching type-checking bugs.

Reply:
The parameter object can't be directly named except through the parameter; it's never the same object as e.g. a local variable in the caller.  However, it can be aliased in the same way that almost any C++ object technically can, e.g. by escaping `this` during construction.  We've generally been assuming that that's not something we need to worry about, at least from the perspective of "well, we're obviously not going to be enforce exclusivity on types that don't follow basic local-reasoning principles".

Reply:
It occurs to me that we probably need special logic when consuming parameters like this.  They probably *should* be consumable as long as the type is movable; we should compile consumption as a C++ move and then enforce use restrictions on the moved-from value just like we would for a Swift value, even though technically the object is still there.